### PR TITLE
[prosemirror-view] update argument type to include marks in nodeViews

### PR DESCRIPTION
**Pull request by BrianHung**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/prosemirror-view/index.d.ts
Comment (ifiokjr) [inline types/prosemirror-view/index.d.ts:571]: Perhaps make this generic.

```suggestion
      nodeOrMark: ProsemirrorNode<S> | Mark<S>,
```

**Comment by ifiokjr** [comment]
This is definitely a breaking change for TypeScript users interacting with `nodeViews`. 

I'd be more confident if you added a test for the new behaviour.

**Review by ifiokjr — CHANGES_REQUESTED** [review]
Since this is a breaking change for a lot of users can you add some tests.

File: types/prosemirror-view/index.d.ts
Comment (BrianHung) [reply types/prosemirror-view/index.d.ts:571]: Making the `mark` type generic is good, though the `nodeOrMark` name is a bit clunky. I think a better solution would be to separate the two entirely by creating a type definition for `MarkView`, so

```ts
  nodeViews?: {
    [name: string]: (
    (
      node: ProsemirrorNode<S>,
      view: EditorView<S>,
      getPos: (() => number),
      decorations: Decoration[]
    ) => NodeView<S>) |
   (
      mark: Mark<S>,
      view: EditorView<S>,
      getPos: boolean,
      decorations: Decoration[]
   ) => MarkView<S>
 )
  } | null;
```

This is more correct, since there will never be a custom NodeView with getPos being a boolean, and a custom Markview with getPos being a function.

Comment (ifiokjr) [reply types/prosemirror-view/index.d.ts:571]: I think union declarations of function properties can cause problems. It's been a while since I tried to use one so it might no longer be an issue in newer versions of TypeScript. 

Last time I checked the way that TypeScript merges the methods is to create an intersection of each positional argument. The type declared would become:

```ts
{
  nodeViews?: {
    [key: string]: (
      arg: ProsemirrorNode & Mark,
      view: EditorView,
      getPos: ((() => number) & false) | ((() => number) & true),
      decorations: Decoration[],
    ) => NodeView<S> & MarkView<S>;
  };
}
```

If this is the case, the automatic intersection types would make it unusable.

I really like the idea of adding a `MarkView` type.

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]
Please add tests for this change

File: types/prosemirror-view/index.d.ts
Comment (BrianHung) [reply types/prosemirror-view/index.d.ts]: I added a `MarkView` type which seems to work. However, testing that types worked correctly required Typescript 3.8.

File: types/prosemirror-view/package.json
Comment (danger-public) [inline types/prosemirror-view/package.json:1]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Review by ifiokjr — CHANGES_REQUESTED** [review]
Really nice work here. I've requested one or two changes for you.

File: types/prosemirror-view/index.d.ts
Comment (BrianHung) [reply types/prosemirror-view/index.d.ts:736]: https://prosemirror.net/docs/ref/#view.NodeView
second paragraph

File: types/prosemirror-view/ts3.8/index.d.ts
Comment (BrianHung) [reply types/prosemirror-view/ts3.8/index.d.ts:569]: Hm, `getPos` is the argument name given in https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews, but I see your point of view. Which one would you prefer more?

Comment (ifiokjr) [reply types/prosemirror-view/ts3.8/index.d.ts:569]: I personally prefer `inline`. At a glance it’s easier to understand what’s happening.

File: types/prosemirror-view/package.json
Comment (BrianHung) [reply types/prosemirror-view/package.json:1]: Where do you see non-spaces being used for new or changed types?

File: types/prosemirror-view/index.d.ts
Comment (BrianHung) [reply types/prosemirror-view/index.d.ts:578]: Can you help me check if the union type is backwards compatible with lower versions? I'm actually not sure if it's the union type in `index.d.ts`, or the way `$ExpectError` should placed within `prosemirror-view-tests.ts`.

The reason why I thought it was 3.8 is because when I ran `npm run lint prosemirror-view`, I got the following
```
ERROR: 118:3  expect  Compile error in typescript@3.7 but not in typescript@3.8.
Fix with a comment '// Minimum TypeScript Version: 3.8' just under the header.
Type '{ markview: (mark: model.Node<any>, view: view.EditorView<any>, getPos: boolean, decorations: view.Decoration<{ [key: string]: any; }>[]) => any; nodeview: (node: model.Mark<any>, view: view.EditorView<...>, getPos: () => number, decorations: view.Decoration<...>[]) => any; }' is not assignable to type '{ [name: string]: ((node: ProsemirrorNode<any>, view: EditorView<any>, getPos: () => number, decorations: Decoration<{ [key: string]: any; }>[]) => NodeView<any>) | ((mark: Mark<...>, view: EditorView<...>, getPos: boolean, decorations: Decoration<...>[]) => MarkView<...>); }'.
  Property 'markview' is incompatible with index signature.
    Type '(mark: model.Node<any>, view: view.EditorView<any>, getPos: boolean, decorations: view.Decoration<{ [key: string]: any; }>[]) => any' is not assignable to type '((node: ProsemirrorNode<any>, view: EditorView<any>, getPos: () => number, decorations: Decoration<{ [key: string]: any; }>[]) => NodeView<any>) | ((mark: Mark<...>, view: EditorView<...>, getPos: boolean, decorations: Decoration<...>[]) => MarkView<...>)'.
      Type '(mark: model.Node<any>, view: view.EditorView<any>, getPos: boolean, decorations: view.Decoration<{ [key: string]: any; }>[]) => any' is not assignable to type '(node: ProsemirrorNode<any>, view: EditorView<any>, getPos: () => number, decorations: Decoration<{ [key: string]: any; }>[]) => NodeView<any>'.
        Types of parameters 'getPos' and 'getPos' are incompatible.
          Type '() => number' is not assignable to type 'boolean'.
```

but I can't pinpoint in the 3.8 release notes (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html) any mention of union type.

**Comment by amcasey** [comment]
Closing as abandoned.  Please ping if you're still working on this.

**Comment by BrianHung** [comment]
@amcasey 

I was trying to get maintainer feedback before pushing my next commit here: https://github.com/BrianHung/DefinitelyTyped/commit/d9d4219e1631f7002f7f45111e32574563cd31c9, mostly with regards to this comment: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46350#discussion_r478873253. 

Could someone more proficient than me in Typescript explain whether the introduction of a union type into the custom nodeview is an issue with _types_, or just an issue with the _testcase_ itself?

File: types/prosemirror-view/index.d.ts
Comment (amcasey) [reply types/prosemirror-view/index.d.ts:578]: I believe the normal pattern would be to put the _older_ version in a folder and keep the newest version in the root.  At least, I assume that's why dtslint complains about `typesVersions` specification.

File: types/prosemirror-view/package.json
Comment (amcasey) [reply types/prosemirror-view/package.json:1]: That's actually a bot.  Bizarrely, it looks like there are trailing tabs on the lines of this file.

File: types/prosemirror-view/index.d.ts
Comment (amcasey) [reply types/prosemirror-view/index.d.ts:578]: There were a bunch of union changes in the 3.8/3.9 timeframe, but I didn't track them closely enough to be able to tell you which affected your change.

Comment (BrianHung) [reply types/prosemirror-view/index.d.ts:578]: Hm, does the latest commit (https://github.com/BrianHung/DefinitelyTyped/commit/d9d4219e1631f7002f7f45111e32574563cd31c9) "put the older version in a folder and keep the newest version in the root"? 

The only difference between this commit and the one we're commenting on is that I added backwards-compatible tests for 3.1 and 3.8; I did not change the types themselves.

`npm run lint prosemirror-view` and `npm test` run without error so it seems fine? I'm still fairly new to testing with Typescript.

**Comment by amcasey** [comment]
@BrianHung, this type seems like it's going to be hard to consume:
```ts
{
    [name: string]: (
      ((
        node: ProsemirrorNode<S>,
        view: EditorView<S>,
        getPos: (() => number),
        decorations: Decoration[])
        => NodeView<S>) |
      ((
        mark: Mark<S>,
        view: EditorView<S>,
        inline: boolean,
        decorations: Decoration[]
      ) => MarkView<S>)
    )
  }
```

Suppose I have an instance called `x`.  Then, for any `myProp`, `x.myProp` is a function with one of two incompatible signatures?  How does the consumer decide which calling pattern to follow?

File: types/prosemirror-view/index.d.ts
Comment (amcasey) [reply types/prosemirror-view/index.d.ts:578]: It would be easier for me to review if you pushed changes here (which I assume you weren't doing because the PR was closed).

**Comment by BrianHung** [comment]
@amcasey 

> Then, for any myProp, x.myProp is a function with one of two incompatible signatures?

Do you mean a function with one of two compatible signatures? Yeah that's the restriction on CustomNodeView types defined in the text of https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews. The original prosemirror author said [here](https://github.com/ProseMirror/prosemirror-view/pull/78#issuecomment-664022408):
> here is never actually a function that accepts these union types — rather, the type of the function is, depending on whether it handles a node or a mark, one of two specific types.

---

> How does the consumer decide which calling pattern to follow?

The consumer, the person writing a library around prosemirror (?), could either know because CustomNodeView type is hard-coded like in the test case
```js
// TestMarkView implements MarkView, and TestNodeView implements NodeView
    markview: (mark: model.Mark, view: view.EditorView, getPos: boolean, decorations: view.Decoration[]) => new TestMarkView(mark, view, getPos, decorations),
    nodeview: (node: model.Node, view: view.EditorView, getPos: () => number, decorations: view.Decoration[]) => new TestNodeView(node, view, getPos, decorations),
```
or distinguish whether the CustomNodeView is a "mark" or "node" in an auxillary attribute, and then we can use that in an "if" statement. 

For example, this is how I'm using it with extension wrappers which implement CustomNodeViews and also have a "type" which distinguishes whether its a "mark" or a "node".

```js
  buildNodeViews(): {
    const reduce = (NodeViews, extension: Node | Mark) => {
      if (extension.customNodeView === undefined) return NodeViews;
      if (extension.type === "mark") {
          return {...NodeViews, [extension.name]: (node: Node, view, getPos: () => number, decorations) => extension.customNodeView({extension, node, view, getPos, decorations})};
      } else {
          return {...NodeViews, [extension.name]: (mark: Mark, view, getPos: boolean, decorations) => extension.customMarkView({extension, node, view, getPos, decorations})};
      }
    }
    const extensions = this.options.extensions.filter(ext => (ext instanceof Node || ext instanceof Mark) && ext.customNodeView)
    return extensions.reduce(reduce, {})
  }
```

File: types/prosemirror-view/index.d.ts
Comment (amcasey) [reply types/prosemirror-view/index.d.ts:578]: I didn't see a new commit to the PR, so I looked at the branch in your fork.  It looks like there's still a ts3.8 folder, so it's not in the structure I expected.  I thought the code for the newest version of TypeScript would go in the root directory (`types/prosemirror-view`).

**Comment by amcasey** [comment]
@BrianHung 

> Do you mean a function with one of two compatible signatures? 

No, I meant incompatible - it seems like `ProsemirrorNode<S>` is not assignable to `Mark<S>` and vice versa.

By "a consumer", I meant code that is getting/reading/retrieving a property of that type.  The value it gets for the property will have a union type, which means it won't be callable unless the consumer some knows which of the two possible types it actually is and uses a type assertion.

If, for the purposes of your library, it is only ever necessary to assign values to those properties, and not to read them, then the union is fine.

**Comment by BrianHung** [comment]
> No, I meant incompatible - it seems like ProsemirrorNode\<S> is not assignable to Mark\<S> and vice versa.

Yeah that's ok. They shouldn't be assignable to each other since they have different APIs, and do not subclass each other:

https://prosemirror.net/docs/ref/#model.Node
https://prosemirror.net/docs/ref/#model.Mark

---

> The value it gets for the property will have a union type, which means it won't be callable unless the consumer some knows which of the two possible types it actually is and uses a type assertion.
> If, for the purposes of your library, it is only ever necessary to assign values to those properties, and not to read them, then the union is fine.

This is fine too I believe. The code that calls the NodeView lambda function we're type annotating is given here https://github.com/ProseMirror/prosemirror-view/blob/master/src/viewdesc.js#L521-L527 and https://github.com/ProseMirror/prosemirror-view/blob/master/src/viewdesc.js#L578-L609. 

The `CustomNodeViewDesc` corresponds to the `NodeView` types here https://github.com/DefinitelyTyped/DefinitelyTyped/blob/04a445ec63ad8dd1c3c47897eaaa31a02bc73d35/types/prosemirror-view/index.d.ts#L657 and `CustomMarkViewDesc` corresponds to the new `MarkView` type that I've added. 

So the consumer by implementing `CustomNodeViewDesc` or `CustomMarkViewDesc` should know which of the two possible types they're using.

**Comment by amcasey** [comment]
@BrianHung can you please add a test showing how that would work?

**Comment by amcasey** [comment]
Before I could sign off, I'd need to see the TS version directories straightened out and the CI tests passing.

**Pull request closed without merging**
